projects/evm-compatible-bridge/federator.md

```diff
@@ -0,0 +1,300 @@
+  - If the token is native to the Hathor chain, the equivalent token is already burned and we need to unlock the tokens on Hathor.
+  - The loop will have to use the coordinator service to send the amount of tokens requested from the MultiSig wallet to the destination address.
+- A request to support a new token native to the EVM chain.
+  - We need to check if this is a duplicate request
+  - We need to use the coordinator service to create of the equivalent token on Hathor.
+  - We need to use the bridge smart contract to save the Hathor equivalent token uid once the token is created.
+- Confirmation that an operation is complete
+  - This way we can clean any state pertaining to this crossing of tokens and mark it as complete.
+  - If the crossing was originated in Hathor, we need to send a request to the coordinator service so it knows the crossing is over.
+
+### Polling loop
+
+We will pool for events, which are requests for voting, and vote on them.
+If the event requires a vote on the smart contract, we will send a transaction to the EVM chain with the vote.
+If the event requires a vote on the MultiSig wallet, we will send a message to the coordinator service with a signature.
+
+The federator will be responsible to check the data on the event, it will have its own headless wallet to check the information from the Hathor network and it can read the data from the EVM chain.
+
+## Coordinator service
+
+The coordinator service will serve the function of the smart contract in the Hathor side.
+It will listen for transactions on the MultiSig wallet, keep a comprehensive persistent database so we do not process the same transaction twice and will gather signatures from the federation.
+It will be started with the MultiSig wallet config but it will not be a participant of the wallet.
+
+The service will download the history of transactions of the MultiSig wallet and process all transactions.
+If a transaction was already processed, the federation contract should have it in its state, so we can safely ignore it.
+
+The coordinator service will provide a REST API for the federators and admins to interact with it.
+
+### Processing Hathor transactions
+
+When first starting the process it will iterate over all transactions in the MultiSig wallet and process them.
+After this, each new message will be enqueued and processed in order of arrival.
+Obs: The headless wallet offers a message queue extension to enqueue events e.g. new transactions.
+
+First we need to check the transaction is compliant with the request format if not, we can save it as processed and continue.
+If the transaction is already on tha database we can skip it, if not we need to check the federation contract to see if it was already processed.
+Obs: This read operation on the smart contract does not require a transaction to be sent, we can read the state from the EVM chain with 0 cost.
```

Comment:
Each request to cross tokens will have an unique id, so we can check if it has already been fulfilled (to avoid double crossing).
When a token has crossed the bridge we should mark this id as 'fulfilled', this can be a simple map of id to boolean (fulfilled or not)

Reply:
How and when is this id created?

Reply:
The transaction id of Hathor is the id of the request to cross the token from Hathor to EVM.
The combination of block hash and transaction hash on the EVM is the id of the request to cross the token from EVM to Hathor.
To make these more uniform we can hash the data so we have a constant size string of bytes to use as id.

Reply:
The transaction id is the hash of the information on the transaction ([ref](https://github.com/onepercentio/tokenbridge/blob/master/bridge/contracts/Federation.sol#L109)).
It should be slightly different since we are required to id requests on Hathor and on solidity but I believe we can achieve a similar result with 2 similar methods:

```solidity
// EVM -> Hathor
function getEVMTransactionId(
        address originalTokenAddress,
        bytes25 receiverAddress,
        uint64 amount, // Hathor value (the least granular) goes up to 8 bytes
        string memory symbol,
        bytes32 blockHash,
        bytes32 transactionHash,
        uint32 logIndex
) public pure returns(bytes32) {
    return keccak256(abi.encodePacked(originalTokenAddress, receiver, amount, symbol, blockHash, transactionHash, logIndex));
}
    
// Hathor -> EVM
function getHTRTransactionId(
        bytes25 originalTokenAddress,
        address receiverAddress,
        uint64 amount, // Hathor value (the least granular) goes up to 8 bytes
        string memory symbol,
        bytes64 tokenUid,
        bytes64 txId
) public pure returns(bytes32) {
    return keccak256(abi.encodePacked(originalTokenAddress, receiver, amount, symbol, tokenUid, txId));
}
```

What do you think?

Reply:
@renankruger what do you think?